x/concentrated-liquidity/internal/math/tick_test.go

```diff
@@ -1,68 +1,242 @@
+			expectedError:      types.ExponentAtPriceOneError{ProvidedExponentAtPriceOne: types.ExponentAtPriceOneMin.Sub(sdk.OneInt()), PrecisionValueAtPriceOneMin: types.ExponentAtPriceOneMin, PrecisionValueAtPriceOneMax: types.ExponentAtPriceOneMax},
+		},
+		"error: exponentAtPriceOne greater than maximum": {
+			tickIndex:          sdk.NewInt(100),
+			exponentAtPriceOne: types.ExponentAtPriceOneMax.Add(sdk.OneInt()),
+			expectedError:      types.ExponentAtPriceOneError{ProvidedExponentAtPriceOne: types.ExponentAtPriceOneMax.Add(sdk.OneInt()), PrecisionValueAtPriceOneMin: types.ExponentAtPriceOneMin, PrecisionValueAtPriceOneMax: types.ExponentAtPriceOneMax},
+		},
+		"random": {
+			tickIndex:          sdk.NewInt(-9111000000),
+			exponentAtPriceOne: sdk.NewInt(-8),
+			expectedPrice:      sdk.MustNewDecFromStr("0.000000000088900000"),
 		},
 	}
 
 	for name, tc := range testCases {
 		tc := tc
 
 		suite.Run(name, func() {
-			sqrtPrice, err := math.TickToSqrtPrice(tc.tickIndex)
+			sqrtPrice, err := math.TickToPrice(tc.tickIndex, tc.exponentAtPriceOne)
+			if tc.expectedError != nil {
+				suite.Require().Error(err)
+				suite.Require().Equal(tc.expectedError.Error(), err.Error())
+				return
+			}
 			suite.Require().NoError(err)
-			suite.Require().Equal(tc.sqrtPriceExpected, sqrtPrice.String())
+			suite.Require().Equal(tc.expectedPrice.String(), sqrtPrice.String())
 
 		})
 	}
 }
 
 func (suite *ConcentratedMathTestSuite) TestPriceToTick() {
-	testCases := []struct {
-		name         string
-		price        sdk.Dec
-		tickExpected string
+	testCases := map[string]struct {
+		price              sdk.Dec
+		exponentAtPriceOne sdk.Int
+		tickExpected       string
+		expectedError      error
 	}{
-		{
-			"happy path",
-			sdk.NewDec(5000),
-			"85176",
+		"50000 to tick with -4 k at price one": {
+			price:              sdk.NewDec(50000),
+			exponentAtPriceOne: sdk.NewInt(-4),
+			tickExpected:       "400000",
+		},
+		"5.01 to tick with -2 k at price one": {
+			price:              sdk.MustNewDecFromStr("5.01"),
+			exponentAtPriceOne: sdk.NewInt(-2),
+			tickExpected:       "401",
+		},
+		"50000.01 to tick with -6 k at price one": {
+			price:              sdk.MustNewDecFromStr("50000.01"),
+			exponentAtPriceOne: sdk.NewInt(-6),
+			tickExpected:       "40000001",
+		},
+		"0.090000889 to tick with -8 k at price one": {
+			price:              sdk.MustNewDecFromStr("0.090000889"),
+			exponentAtPriceOne: sdk.NewInt(-8),
+			tickExpected:       "-999991110",
+		},
+		"0.9998 to tick with -4 k at price one": {
+			price:              sdk.MustNewDecFromStr("0.9998"),
+			exponentAtPriceOne: sdk.NewInt(-4),
+			tickExpected:       "-20",
+		},
+		"53030.10 to tick with -5 k at price one": {
+			price:              sdk.MustNewDecFromStr("53030.1"),
+			exponentAtPriceOne: sdk.NewInt(-5),
+			tickExpected:       "4030301",
+		},
+		"max spot price and minimum exponentAtPriceOne": {
+			price:              MaxSpotPrice,
```

Comment:
Let's add tests with `MaxSpotPrice` + 1, `MinSpotPrice` and `MinSpotPrice` - 1 to test the edge cases please

Reply:
Added MaxSpotPrice+1 and MinSpotPrice https://github.com/osmosis-labs/osmosis/pull/3874/commits/00c430bab18e8ac9d4776180d677ca3b65aaaec5

Couldn't add MinSpotPrice - 1 since that is already checked by "price must be positive" https://github.com/osmosis-labs/osmosis/blob/00c430bab18e8ac9d4776180d677ca3b65aaaec5/x/concentrated-liquidity/internal/math/tick_test.go#L213-L217

This is because min spot price is the minimum a sdk.Dec could possibly be, and any smaller it would be zero